### PR TITLE
build: update to Go 1.21.0

### PR DESCRIPTION
**Pull request by Roasbeef**

### CODE REVIEW

**Comment by guggero** [comment]
The RPC and linter containers also need to switch to `-alpine` (or `-bullseye` where we do some `apt install`).

**Comment by Roasbeef** [comment]
Pushed up a new version that's moved to bookworm.

**Comment by guggero** [comment]
The RPC compilation docker file needs a change too, as we have hard-coded versions in there (to make sure the compiled/formatted output is deterministic). See https://github.com/lightningnetwork/lnd/pull/7944#discussion_r1312706437 for more info.

File: Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # /make/builder.Dockerfile
 # /taprpc/Dockerfile
 # /tools/Dockerfile
-FROM golang:1.20.5-alpine as builder
+FROM golang:1.21.0-bookworm as builder
```
Comment (guggero) [inline Dockerfile:7]: We can't just switch from `-alpine` (completely different Linux distro with a different package manager) over to Debian. And the Alpine ones are more stable in terms of us just being able to bump the version numbers, compared to Debian. So I'd rather do it the other way, switch from Debian to Alpine (but that requires deeper changes as well, so maybe let's defer that).

File: dev.Dockerfile [hunk 2]
```diff
@@ -15,12 +15,12 @@ ENV CGO_ENABLED=0
 RUN make install
 
 # FINAL IMAGE
-FROM alpine as final
+FROM bookworm as final
```
Comment (guggero) [inline dev.Dockerfile:18]: Same here, we want to keep the slim alpine base image, as the Debian is much larger.

File: taprpc/Dockerfile [hunk 3]
```diff
@@ -4,7 +4,7 @@
 # /dev.Dockerfile
 # /make/builder.Dockerfile
 # /tools/Dockerfile
-FROM golang:1.20.5-buster
+FROM golang:1.21.0-bookworm
```
Comment (guggero) [inline taprpc/Dockerfile:7]: Need to bump the protobuf compiler and formatter below too, see comment above.

**Review by guggero — CHANGES_REQUESTED** [review]
Almost there!
Linter seems to have been a flake, passes now.

**Review by guggero — APPROVED** [review]
Nice, LGTM :tada:

File: Dockerfile (hunk 1 above)
Comment (jharveyb) [reply Dockerfile:7]: This switch of distro is still in here?

Comment (guggero) [reply Dockerfile:7]: Oh yes, I missed this one in my review.

**Pull request merged**